File: Grammars/Phonotactics/Phonotactics.vb
Imports System.Xml.Serialization
Imports Utility
Imports System.Text
Imports System.Globalization
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports YamlDotNet.Serialization

Namespace Phonotactics

    Public Enum CaseType
        None
        Proper
        Upper
        Lower
    End Enum

    <System.Xml.Serialization.XmlRoot("Phonotactics", IsNullable:=False)>
    Partial Public Class PhonotacticsGrammar
        Inherits BaseGrammar

        Public Shared Function GenerateGrammar(ByVal name As String,
                                               ByVal description As String,
                                               ByVal author As String,
                                               ByVal tags As BindingList(Of String),
                                               ByVal supportsMaxLength As Boolean,
                                               ByVal sampleFile As String,
                                               ByVal definitionFile As String,
                                               ByVal maxWeight As Int32) As PhonotacticsGrammar
            Dim grammar As New PhonotacticsGrammar
            Dim samples As New List(Of String)(IO.File.ReadAllLines(sampleFile))
            Dim definitions As New List(Of String)(IO.File.ReadAllLines(definitionFile))

            grammar.Name = name
            grammar.Author = author
            grammar.Tags = tags
            grammar.Description = description
            grammar.SupportsMaxLength = supportsMaxLength

            For Each definition As String In definitions
                Dim key As Char = definition(0)
                Dim newDef As New Phonotactics.Definition
                newDef.Delimiter = ","
                newDef.Key = key
                newDef.Value = definition.Substring(2)
                DirectCast(grammar, Phonotactics.PhonotacticsGrammar).Definitions.Add(newDef)
            Next

            For Each sample As String In samples
                If Not String.IsNullOrWhiteSpace(sample) Then
                    Dim item As String = sample
                    Dim result(item.Length) As String
                    Dim pattern As String
                    For Each definition As Phonotactics.Definition In DirectCast(grammar, Phonotactics.PhonotacticsGrammar).Definitions
                        For Each value As String In definition.ValueList
                            Dim index As Int32 = CultureInfo.CurrentCulture.CompareInfo.IndexOf(item, value, CompareOptions.IgnoreCase)
                            Do While index >= 0
                                result(index) = definition.Key
                                index = CultureInfo.CurrentCulture.CompareInfo.IndexOf(item, value, index + 1, CompareOptions.IgnoreCase)
                            Loop
                        Next
                    Next
                    pattern = Join(result, String.Empty)
                    pattern = pattern.Trim
                    If DirectCast(grammar, Phonotactics.PhonotacticsGrammar).Patterns.FirstOrDefault(Function(p As Phonotactics.Pattern) p.Value = pattern) Is Nothing Then
                        DirectCast(grammar, Phonotactics.PhonotacticsGrammar).Patterns.Add(New Phonotactics.Pattern(pattern))
                    Else
                        DirectCast(grammar, Phonotactics.PhonotacticsGrammar).Patterns.FirstOrDefault(Function(p As Phonotactics.Pattern) p.Value = pattern).Weight += 1
                    End If
                End If
            Next

            Return grammar
        End Function

        Private Class PatternPart
            Public Sub New()
            End Sub

            Public Sub New(ByVal key As String,
                           ByVal [optional] As Boolean)
                Me.Key = key
                Me.Optional = [optional]
            End Sub

            Public Property Key As String
            Public Property [Optional] As Boolean

        End Class

        <XmlElement("case")>
        Public Property [Case]() As CaseType
        <XmlArray("definitions")>
        <XmlArrayItem("item", IsNullable:=False)>
        Public Property Definitions() As New ObservableCollection(Of Definition)
        <XmlArray("patterns")>
        <XmlArrayItem("item", IsNullable:=False)>
        Public Property Patterns() As New ObservableCollection(Of Pattern)

        Public Overrides Function Analyze() As String
            Const tableRow As String = "<tr>{0}</tr>"
            Const tableCell As String = "<td>{0}</td>"

            Dim html As String = My.Resources.PhonotacticsGrammarAnalysis
            Dim parameters As New StringBuilder

            html = html.Replace("[Name]", Me.Name)
            html = html.Replace("[Description]", Me.Description)
            html = html.Replace("[Author]", Me.Author)
            html = html.Replace("[Category]", Me.Category)
            html = html.Replace("[ParameterCount]", Me.Parameters.Count.ToString)
            html = html.Replace("[DefinitionCount]", Me.Definitions.Count.ToString("#,##0"))
            html = html.Replace("[PatternCount]", Me.Patterns.Count.ToString("#,##0"))
            html = html.Replace("[XMLLength]", Split(Me.ToString, Environment.NewLine).Count.ToString("#,##00"))

            For Each parameter As Parameter In Me.Parameters
                Dim row As New StringBuilder
                row.AppendFormatLine(tableCell, parameter.Name)
                parameters.AppendFormatLine(tableRow, row.ToString)
            Next

            html = html.Replace("[Parameters]", parameters.ToString)

            Return html
        End Function

        Protected Overrides Function GenerateName() As String
            Dim pattern As String = SelectPattern()
            Dim patternParts As List(Of PatternPart) = ParsePattern(pattern)
            Dim name As String = String.Empty

            For Each part As PatternPart In patternParts
                If Not part.Optional OrElse (part.Optional AndAlso Random.Next(100) > 50) Then
                    name &= GetValue(part.Key)
                End If
            Next

            Select Case [Case]
                Case CaseType.Lower : name = name.ToLower
                Case CaseType.Upper : name = name.ToUpper
                Case CaseType.Proper : name = StrConv(name, VbStrConv.ProperCase)
            End Select

            Return name
        End Function

        Private Function SelectPattern() As String
            Dim totalWeight As Int32 = Patterns.Sum(Function(p As Pattern) Math.Max(p.Weight, 1))
            Dim selected As Int32 = (BaseGrammar.Random.Next() Mod totalWeight) + 1
            Dim currentWeight As Int32 = 0
            For Each pattern As Pattern In Patterns
                currentWeight += pattern.Weight
                If currentWeight >= selected Then Return pattern.Value
            Next
            Return String.Empty
        End Function

        Private Function ParsePattern(ByVal pattern As String) As List(Of PatternPart)
            Const OPTIONAL_START As Char = "("c
            Const OPTIONAL_END As Char = ")"c

            Dim [optional] As Boolean = False
            Dim current As Char = Nothing
            Dim value As New List(Of PatternPart)

            For Each c As Char In pattern
                Select Case c
                    Case OPTIONAL_START
                        [optional] = True
                    Case OPTIONAL_END
                    Case Else
                        value.Add(New PatternPart(c, [optional]))
                        [optional] = False
                End Select
            Next

            Return value
        End Function

        Private Function GetValue(ByVal key As String) As String
            Dim definition As Definition = Definitions.FirstOrDefault(Function(d As Definition) d.Key = key)
            If definition IsNot Nothing Then
                Dim values As List(Of String) = definition.ValueList
                Dim index As Int32 = BaseGrammar.Random.Next Mod values.Count
                Return values(index)
            Else
                Return key
            End If
        End Function

    End Class

    Partial Public Class Definition

        Public Sub New()
        End Sub

        Public Sub New(ByVal key As String,
                       ByVal delimiter As Char,
                       ByVal value As String)
            Me.Key = key
            Me.Delimiter = delimiter
            Me.Value = value
        End Sub

        Private _valueList As List(Of String)

        <XmlAttribute("key")>
        Public Property Key() As String
        <XmlAttribute("delimiter")>
        Public Property Delimiter() As String
        <XmlText()>
        Public Property Value() As String

        <XmlIgnore>
        <YamlIgnore>
        Friend ReadOnly Property ValueList() As List(Of String)
            Get
                If _valueList Is Nothing Then
                    _valueList = New List(Of String)
                    _valueList.AddRange(Split(Value, Delimiter))
                End If
                Return _valueList
            End Get
        End Property

    End Class

    Partial Public Class Pattern
        Public Sub New()
        End Sub

        Public Sub New(ByVal value As String)
            Me.Value = value
        End Sub

        <XmlText()>
        Public Property Value() As String
        <XmlAttribute("weight")>
        Public Property Weight() As Int32 = 1
    End Class
End Namespace

File: Grammars/Phonotactics/Phonotactics.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;
using System.Xml.Serialization;
using Microsoft.VisualBasic;
using Utility;
using YamlDotNet.Serialization;

namespace Grammars.Phonotactics
{
    public enum CaseType
    {
        None,
        Proper,
        Upper,
        Lower
    }

    [XmlRoot("Phonotactics", IsNullable = false)]
    public class PhonotacticsGrammar : BaseGrammar
    {
        [XmlArray("definitions")]
        [XmlArrayItem("item", IsNullable = false)]
        private ObservableCollection<Definition> _Definitions = new ObservableCollection<Definition>();

        [XmlArray("patterns")]
        [XmlArrayItem("item", IsNullable = false)]
        private ObservableCollection<Pattern> _Patterns = new ObservableCollection<Pattern>();

        [XmlElement("case")]
        public CaseType Case { get; set; }

        private ObservableCollection<Definition> Definitions
        {
            get { return _Definitions; }
            set { _Definitions = value; }
        }

        private ObservableCollection<Pattern> Patterns
        {
            get { return _Patterns; }
            set { _Patterns = value; }
        }

        public static PhonotacticsGrammar GenerateGrammar(
            string name,
            string description,
            string author,
            BindingList<string> tags,
            bool supportsMaxLength,
            string sampleFile,
            string definitionFile,
            int maxWeight)
        {
            var grammar = new PhonotacticsGrammar();
            var samples = new List<string>(File.ReadAllLines(sampleFile));
            var definitions = new List<string>(File.ReadAllLines(definitionFile));

            grammar.Name = name;
            grammar.Author = author;
            grammar.Tags = tags;
            grammar.Description = description;
            grammar.SupportsMaxLength = supportsMaxLength;

            foreach (var definition in definitions)
            {
                var key = definition[0];
                var newDef = new Definition
                {
                    Delimiter = ",",
                    Key = key.ToString(),
                    Value = definition.Substring(2)
                };
                grammar.Definitions.Add(newDef);
            }

            foreach (var sample in samples)
            {
                if (!string.IsNullOrWhiteSpace(sample))
                {
                    var item = sample;
                    var result = new string[item.Length + 1];
                    var pattern = "";
                    foreach (var definition in grammar.Definitions)
                    {
                        foreach (var value in definition.ValueList)
                        {
                            var index = CultureInfo.CurrentCulture.CompareInfo.IndexOf(
                                item,
                                value,
                                CompareOptions.IgnoreCase);
                            while (index >= 0)
                            {
                                result[index] = definition.Key;
                                index = CultureInfo.CurrentCulture.CompareInfo.IndexOf(
                                    item,
                                    value,
                                    index + 1,
                                    CompareOptions.IgnoreCase);
                            }
                        }
                    }
                    pattern = string.Join(string.Empty, result);
                    pattern = pattern.Trim();
                    if (grammar.Patterns.FirstOrDefault(p => p.Value == pattern) == null)
                    {
                        grammar.Patterns.Add(new Pattern(pattern));
                    }
                    else
                    {
                        grammar.Patterns.FirstOrDefault(p => p.Value == pattern).Weight++;
                    }
                }
            }

            return grammar;
        }

        public override string Analyze()
        {
            const string tableRow = "<tr>{0}</tr>";
            const string tableCell = "<td>{0}</td>";

            var html = Resources.PhonotacticsGrammarAnalysis;
            var parameters = new StringBuilder();

            html = html.Replace("[Name]", Name);
            html = html.Replace("[Description]", Description);
            html = html.Replace("[Author]", Author);
            html = html.Replace("[Category]", Category);
            html = html.Replace("[ParameterCount]", Parameters.Count.ToString());
            html = html.Replace("[DefinitionCount]", Definitions.Count.ToString("#,##0"));
            html = html.Replace("[PatternCount]", Patterns.Count.ToString("#,##0"));
            html = html.Replace(
                "[XMLLength]",
                Strings.Split(ToString(), Environment.NewLine).Length.ToString("#,##00"));

            foreach (var parameter in Parameters)
            {
                var row = new StringBuilder();
                row.AppendFormatLine(tableCell, parameter.Name);
                parameters.AppendFormatLine(tableRow, row.ToString());
            }

            html = html.Replace("[Parameters]", parameters.ToString());

            return html;
        }

        protected override string GenerateName()
        {
            var pattern = SelectPattern();
            var patternParts = ParsePattern(pattern);
            var name = string.Empty;

            foreach (var part in patternParts)
            {
                if (!part.Optional || (part.Optional && Random.Next(100) > 50))
                {
                    name += GetValue(part.Key);
                }
            }

            switch (Case)
            {
                case CaseType.Lower:
                    name = name.ToLower();
                    break;
                case CaseType.Upper:
                    name = name.ToUpper();
                    break;
                case CaseType.Proper:
                    name = Strings.StrConv(name, VbStrConv.ProperCase);
                    break;
            }

            return name;
        }

        private string SelectPattern()
        {
            var totalWeight = Patterns.Sum(p => Math.Max(p.Weight, 1));
            var selected = Random.Next()%totalWeight + 1;

            var currentWeight = 0;
            foreach (var pattern in Patterns)
            {
                currentWeight += pattern.Weight;
                if (currentWeight >= selected)
                {
                    return pattern.Value;
                }
            }

            return string.Empty;
        }

        private List<PatternPart> ParsePattern(string pattern)
        {
            const char OPTIONAL_START = '(';
            const char OPTIONAL_END = ')';

            var optional = false;
            var value = new List<PatternPart>();

            foreach (var c in pattern)
            {
                switch (c)
                {
                    case OPTIONAL_START:
                        optional = true;
                        break;
                    case OPTIONAL_END:
                        break;
                    default:
                        value.Add(new PatternPart(c.ToString(), optional));
                        optional = false;
                        break;
                }
            }

            return value;
        }

        private string GetValue(string key)
        {
            var definition = Definitions.FirstOrDefault(d => d.Key == key);
            if (definition != null)
            {
                var values = definition.ValueList;
                var index = Convert.ToInt32(Random.Next()%values.Count);
                return values[index];
            }
            return key;
        }

        private class PatternPart
        {
            public PatternPart()
            {
            }

            public PatternPart(string key, bool optional)
            {
                Key = key;
                Optional = optional;
            }

            public string Key { get; }
            public bool Optional { get; }
        }
    }

    public class Definition
    {
        private List<string> _valueList;

        public Definition()
        {
        }

        public Definition(string key, char delimiter, string value)
        {
            Key = key;
            Delimiter = delimiter.ToString();
            Value = value;
        }

        [XmlAttribute("key")]
        public string Key { get; set; }

        [XmlAttribute("delimiter")]
        public string Delimiter { get; set; }

        [XmlText]
        public string Value { get; set; }

        [XmlIgnore]
        [YamlIgnore]
        internal List<string> ValueList
        {
            get
            {
                if (_valueList == null)
                {
                    _valueList = new List<string>();
                    _valueList.AddRange(Value.Split(Delimiter.ToCharArray()[0]));
                }
                return _valueList;
            }
        }
    }

    public class Pattern
    {
        [XmlAttribute("weight")]
        private int _Weight = 1;

        public Pattern()
        {
        }

        public Pattern(string value)
        {
            Value = value;
        }

        [XmlText]
        public string Value { get; set; }

        public int Weight
        {
            get { return _Weight; }
            set { _Weight = value; }
        }
    }
}